pkg/cli/cli_test.go

```diff
@@ -17,120 +17,142 @@ limitations under the License.
+		pluginAV1       = makeAllPlugin(pluginNameA, "v1.0", projectVersions...)
+		pluginAV2       = makeAllPlugin(pluginNameA, "v2.0", projectVersions...)
+		pluginBV1       = makeAllPlugin(pluginNameB, "v1.0", projectVersions...)
+		pluginBV2       = makeAllPlugin(pluginNameB, "v2.0", projectVersions...)
+		allPlugins      = []plugin.Base{pluginAV1, pluginAV2, pluginBV1, pluginBV2}
+	)
+
+	Describe("New", func() {
+
+		Context("with no plugins specified", func() {
+			It("should return a valid CLI", func() {
+				By("setting one plugin")
+				c, err = New(WithDefaultPlugins(pluginAV1), WithPlugins(pluginAV1))
+				Expect(err).NotTo(HaveOccurred())
+				Expect(c).NotTo(BeNil())
+				Expect(c.(*cli).pluginsFromOptions).To(Equal(makeSetByProjVer(pluginAV1)))
+				Expect(c.(*cli).resolvedPlugins).To(Equal([]plugin.Base{pluginAV1}))
+
+				By("setting two plugins with different names and versions")
+				c, err = New(WithDefaultPlugins(pluginAV1), WithPlugins(pluginAV1, pluginBV2))
+				Expect(err).NotTo(HaveOccurred())
+				Expect(c).NotTo(BeNil())
+				Expect(c.(*cli).pluginsFromOptions).To(Equal(makeSetByProjVer(pluginAV1, pluginBV2)))
+				Expect(c.(*cli).resolvedPlugins).To(Equal([]plugin.Base{pluginAV1}))
+
+				By("setting two plugins with the same names and different versions")
+				c, err = New(WithDefaultPlugins(pluginAV1), WithPlugins(pluginAV1, pluginAV2))
+				Expect(err).NotTo(HaveOccurred())
+				Expect(c).NotTo(BeNil())
+				Expect(c.(*cli).pluginsFromOptions).To(Equal(makeSetByProjVer(pluginAV1, pluginAV2)))
+				Expect(c.(*cli).resolvedPlugins).To(Equal([]plugin.Base{pluginAV1}))
+
+				By("setting two plugins with different names and the same version")
+				c, err = New(WithDefaultPlugins(pluginAV1), WithPlugins(pluginAV1, pluginBV1))
+				Expect(err).NotTo(HaveOccurred())
+				Expect(c).NotTo(BeNil())
+				Expect(c.(*cli).pluginsFromOptions).To(Equal(makeSetByProjVer(pluginAV1, pluginBV1)))
+				Expect(c.(*cli).resolvedPlugins).To(Equal([]plugin.Base{pluginAV1}))
+			})
+
+			It("should return an error", func() {
+				By("not setting any plugins or default plugins")
+				_, err = New()
+				Expect(err).To(MatchError(`no plugins for project version "3-alpha"`))
+
+				By("not setting any plugin")
+				_, err = New(WithDefaultPlugins(pluginAV1))
+				Expect(err).To(MatchError(`no plugins for project version "3-alpha"`))
+
+				By("not setting any default plugins")
+				_, err = New(WithPlugins(pluginAV1))
+				Expect(err).To(MatchError(`no default plugins for project version "3-alpha"`))
+
+				By("setting two plugins of the same name and version")
+				_, err = New(WithDefaultPlugins(pluginAV1), WithPlugins(pluginAV1, pluginAV1))
+				Expect(err).To(MatchError(`broken pre-set plugins: two plugins have the same key: "go.example.com/v1.0"`))
+			})
 		})
-	})
 
-	Describe("Check if has duplicate plugins", func() {
-		It("should work successfully when a plugin has many versions", func() {
-			plugins := makePluginsForKeys(
```

Comment:
we should not allow duplicating plugins ( same name/version )

Reply:
This is tested [here](https://github.com/kubernetes-sigs/kubebuilder/pull/1536/files#diff-4c87b5f848c90e8582b2bad56402342fR91).